addons/website/static/src/components/configurator/configurator.js

```diff
@@ -417,6 +417,12 @@ ThemeSelectionScreen.template = 'website.Configurator.ThemeSelectionScreen';
 class Configurator extends Component {
     setup() {
         this.router = useRouter();
+        // Using the back button must update the router state.
+        const backListener = ev => {
+            this.router.location = window.location.pathname;
+        };
+        window.addEventListener("popstate", backListener);
+        onWillDestroy(() => window.removeEventListener("popstate", backListener));
```

Comment:
A bit confused about how it worked before, there were no code handling it and it magically worked? It seems strange to have to do what you do here by yourself (why is the router not automatically updated? :thinking: or at least why is not there an utils / options to do that automatically when using the router?)

Reply:
The configurator used to use the builtin owl router in owl 1, the router was removed from owl as it's a completely separate concern from managing the UI. A *very* simple router was reimplemented within the configurator itself when migrating to owl 2 (basically just a reactive object that will render the app when the route changes, and a bunch of if-elif in the template) and we forgot to implement history management.

Reply:
Ok, but still: should not this code appear on a lower level of the code? I see many `this.router = useRouter();` in this file, why don't we have to do this for all of them? (= why is not listening to `popstate` and updating the router object not implemented inside the router itself?)

@bso-odoo @sdegueldre

Reply:
Because this is the only component that actually reads the router state, the others only use it to navigate (also even if that wasn't the case, you'd still only want one listener because there is only one router). I do agree that this should be implemented inside the router and suggested an implementation strategy higher in the conversation.

I think the reason that Benoit chose not to do that is because the router is not reactive within its constructor (it's made reactive in useRouter) which means that the listeners that it adds would modify the state of the router without going through the reactive layer which would bypass the reactivity when it mutates itself and not notify components, hence my suggestion to construct a reactive in the constructor and return it instead, but admittedly returning an object from a constructor is not good style.

Reply:
Thank you for the summary. And indeed I did not feel comfortable returning another object from the constructor. (I even thought for some time that it was meant as a joke.)

Reply:
If you really don't like it you could also have a static factory method "create/init/makeReactiveRouter" that calls new, makes the object reactive and adds the listeners and returns it, this is a pattern that's closer to what you'd see in other languages like Java or C# but at this point it's just different syntax to do the same thing (while also allowing the constructor to be called directly which is undesirable because it doesn't make a "valid" router object).

Reply:
Thank you, the factory argument convinced me of using the return in the constructor.
I handled the removal of the listener slightly differently because `onWillDestroy` cannot be set up in that context (there is no `currentNode`), so I instead used the `reactive`'s `callback` to find out we left the configurator.
I end up having this silly-looking line though:
```js
        // Enable reactivity.
        reactiveRouter.location;
```
I could make it more discrete like this, but I guess it would be even more confusing:
```js
        return reactiveRouter.location && reactiveRouter;
```

Reply:
Do we actually need to remove the listener? Last I checked the configurator was a self-contained app and navigating out of the configurator would cause a page reload anyway. I kindda just missed that we were not in a component when I wrote my suggestion, and having a cleanup for global listeners in component is just good practice but in this case I don't think it's needed.